### PR TITLE
Read ECID from IdentityDirect on Boot when registered

### PR DESCRIPTION
**Pull request by kevinlind**
## Description

Change the Edge Identity boot sequence to first check if Identity direct is registered and if so, wait for Identity direct shared state to set ECID. This change allows for users who install an application new which uses both Edge Identity and Identity direct extensions to sync the ECIDs on first launch. New sequence is as follows:

1. If ECID is in Edge Identity persistence, load from persistence (after first launch, this should always be the case)
2. Else if ECID is in Identity direct persistence, load from Identity direct persistence
3. Else if Identity direct is registered, exit boot sequence and run again when Identity direct dispatches shared state change event
4. Else, generate new ECID.

Note, requires fix in AEPCore v3.1.1

Internal ticket: AMSDK-11398

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: Sources/IdentityConstants.swift [hunk 1]
```diff
@@ -24,9 +24,22 @@ enum IdentityConstants {
         static let ZERO_ADVERTISING_ID = "00000000-0000-0000-0000-000000000000"
     }
 
-    enum SharedStateKeys {
-        static let CONFIGURATION = "com.adobe.module.configuration"
-        static let IDENTITY_DIRECT = "com.adobe.module.identity"
+    enum SharedState {
```
Comment (kevinlind) [inline Sources/IdentityConstants.swift:27]: Note, I moved some constants around so they are similar to Edge. There will be several line diff just for that change.

File: Sources/IdentityState.swift [hunk 2]
```diff
@@ -42,6 +42,9 @@ class IdentityState {
             if let ecid = identityProperties.getEcidFromDirectIdentityPersistence() {
                 identityProperties.ecid = ecid.ecidString // get ECID from direct extension
                 Log.debug(label: IdentityConstants.LOG_TAG, "IdentityState - Loading ECID from direct Identity extension on bootup '\(ecid)'")
+            } else if isIdentityDirectRegistered(event: event, getSharedState: getSharedState) {
+                Log.debug(label: IdentityConstants.LOG_TAG, "IdentityState - bootup direct Identity extention is registered, waiting for its state change.")
+                return false // If no ECID to migrate but Identity direct is registered, wait for Identity direct shared state
```
Comment (kevinlind) [inline Sources/IdentityState.swift:47]: I'm not handling the case where Identity direct is registered but the shared state doesn't contain an ECID (opted-out case). I'll need to update this.

File: Sources/Identity.swift [hunk 3]
```diff
@@ -36,19 +36,22 @@ import Foundation
...
     public func onUnregistered() {
     }
 
     public func readyForEvent(_ event: Event) -> Bool {
-        return true
+        if state.hasBooted {
+            return true
+        }
+
+        if state.bootupIfReady(event: event, getSharedState: getSharedState(extensionName:event:barrier:)) {
```
Comment (nporter-adbe) [inline Sources/Identity.swift:49]: It is required to pass in the `barrier` version of the getSharedState API? You should be able to just `getSharedState(extensionName:event:)` for the standard shared state.

File: Sources/IdentityState.swift [hunk 4]
```diff
@@ -142,4 +145,19 @@ class IdentityState {
...
         createXDMSharedState(identityProperties.toXdmData(), event)
     }
+
+    /// Check if the Identity direct extension is registered by checking the EventHub's shared state list of registered extensions.
+    /// - Parameters:
+    ///   - event: the tiggering event
+    ///   - getSharedState: function to get shared states from the EventHub
+    /// - Returns: true if the Identity direct extension is registered with the EventHub
+    private func isIdentityDirectRegistered(event: Event, getSharedState: @escaping (_ name: String, _ event: Event?, _ barrier: Bool) -> SharedStateResult?) -> Bool {
+        if let registeredExtensionsWithHub = getSharedState(IdentityConstants.SharedState.Hub.SHARED_OWNER_NAME, event, false)?.value,
```
Comment (emdobrin) [inline Sources/IdentityState.swift:155]: Should we rather pass event nil to fetch latest at the time we do this check as I was discussing offline with Nick?

File: Sources/IdentityState.swift [hunk 5]
```diff
@@ -31,7 +31,7 @@ class IdentityState {
 
     /// Completes init for this Identity extension.
     /// - Returns: True if we should share state after bootup, false otherwise
```
Comment (emdobrin) [inline Sources/IdentityState.swift:33]: It is a bit confusing on when bootupIfReady returns true/false, it looks like we are using it to determine if it booted and also if shared state needs to be updated or not. 
We can consolidate that by doing the following changes:
1. `if hasBooted { return true }`
2. create shared state inside IdentityState when bootup is completed
3. simplify readyForEvent

File: Sources/IdentityState.swift (hunk 4 above)
Comment (kevinlind) [reply Sources/IdentityState.swift]: Yes, I changed this to get the latest shared state.

**Review by nporter-adbe — APPROVED** [review]

**Review by PravinPK — APPROVED** [review]

File: Sources/IdentityState.swift (hunk 5 above)
Comment (kevinlind) [reply Sources/IdentityState.swift]: I've added `createXDMSharedState` as a parameter to `bootupIfReady` and changed the return value to be true/false if the bootup completed successfully. I also added more comments to document the boot up process.

**Review by nporter-adbe — APPROVED** [review]
👍🏻👍🏻

**Pull request merged**